tests/unit/meta/meta.ts

```diff
@@ -0,0 +1,201 @@
+		}
+
+		class TestWidget extends ProjectorMixin(WidgetBase)<any> {
+			render() {
+				return v('div', {
+					innerHTML: 'hello world',
+					key: 'root'
+				});
+			}
+
+			getMeta() {
+				return this.meta(TestMeta);
+			}
+		}
+
+		const div = document.createElement('div');
+
+		const widget = new TestWidget();
+		widget.append(div);
+
+		assert.strictEqual(widget.getMeta(), widget.getMeta());
+	},
+
+	'meta is provided a list of nodes with keys'() {
+		class TestMeta extends MetaBase {
+		}
+
+		class TestWidget extends ProjectorMixin(WidgetBase)<any> {
+			render() {
+				return v('div', {
+					innerHTML: 'hello world',
+					key: 'root'
+				});
+			}
+
+			getMeta() {
+				return this.meta(TestMeta);
+			}
+		}
+
+		const div = document.createElement('div');
+
+		const widget = new TestWidget();
+		widget.append(div);
+		const meta = widget.getMeta();
+
+		assert.isTrue((<any> meta).nodes.size > 0);
+		assert.isNotNull((<any> meta).nodes.get('root'));
+	},
+
+	'meta renders the node if it has to'() {
+		class TestMeta extends MetaBase {
+		}
+
+		let renders = 0;
+
+		class TestWidget extends ProjectorMixin(WidgetBase)<any> {
+			nodes: any;
+
+			render() {
+				renders++;
+
+				this.meta(TestMeta).has('greeting');
+				this.meta(TestMeta).has('name');
+
+				return v('div', {
+					innerHTML: 'hello',
+					key: 'greeting'
+				}, [
+					v('div', {
+						innerHTML: 'world',
+						key: 'name'
+					})
+				]);
+			}
+		}
+
+		const div = document.createElement('div');
+
+		const widget = new TestWidget();
+		widget.append(div);
+
+		resolveRAF();
+
+		assert.strictEqual(renders, 2, 'expected two renders');
+	},
+
+	'multi-step render'() {
+		class TestMeta extends MetaBase {
+		}
+
+		let renders = 0;
+
+		class TestWidget extends ProjectorMixin(WidgetBase)<any> {
+			nodes: any;
+
+			render() {
+				renders++;
+
+				const test = this.meta(TestMeta);
+
+				return v('div', {
+					innerHTML: 'hello',
+					key: 'greeting'
+				}, [
+					test.has('greeting') ? v('div', {
+						innerHTML: 'world',
+						key: 'name'
+					}, [
+						test.has('name') ? v('div', {
+							innerHTML: '!',
+							key: 'exclmation'
+						}) : null
+					]) : null
+				]);
+			}
+		}
+
+		const div = document.createElement('div');
+
+		const widget = new TestWidget();
+		widget.append(div);
+
+		resolveRAF();
+
+		assert.strictEqual(renders, 3, 'expected two renders');
+	},
+
+	'meta throws an error if a required node is not found'() {
+		class TestMeta extends MetaBase {
+		}
+
+		let renders = 0;
+
+		class TestWidget extends ProjectorMixin(WidgetBase)<any> {
+			nodes: any;
+
+			render() {
+				renders++;
+
+				this.meta(TestMeta).has('test');
+
+				return v('div', {
+					innerHTML: 'hello world',
+					key: 'root'
+				});
+			}
+		}
+
+		const div = document.createElement('div');
+
+		const widget = new TestWidget();
+		widget.append(div);
+
+		assert.throws(() => {
+			resolveRAF();
+		});
+	},
+
+	'base meta has no-op functions'() {
```

Comment:
I put it there just for test coverage. I'm not able to make the class abstract because it would require anything extending it to implement those no-op methods. There's probably an Istanbul instruction to ignore methods, though.